pandas/core/internals/blocks.py

```diff
@@ -862,14 +866,21 @@ def setitem(self, indexer, value):
             #  be e.g. a list; see GH#6043
             values[indexer] = value
 
-        # if we are an exact match (ex-broadcasting),
-        # then use the resultant dtype
         elif (
-            len(arr_value.shape)
-            and arr_value.shape[0] == values.shape[0]
-            and arr_value.size == values.size
+            self.is_categorical_astype(arr_value.dtype)
```

Comment:
use ``is_categorical_dtype(arr_value.dtype)``

is_categorical_astype is not necessary here

Reply:
Made the changes, and merged master. All tests passed.

Reply:
Hi @jreback, is the PR good to merge to master?